File: zig/src/candlestick_patterns/patterns/shooting_star.zig
/// Shooting Star: a two-candle bearish reversal pattern.
///
/// Must have:
/// - gap up from the previous candle (real body gap up),
/// - small real body,
/// - long upper shadow,
/// - very short lower shadow.
///
/// Returns:
///     Continuous float in [-100, 0].  More negative = stronger signal.

const cp = @import("../candlestick_patterns.zig");
const operators = @import("fuzzy").operators;

const CandlestickPatterns = cp.CandlestickPatterns;
const CriterionState = cp.CriterionState;
const isRealBodyGapUp = cp.isRealBodyGapUp;
const lowerShadow = cp.lowerShadow;
const realBodyLen = cp.realBodyLen;
const upperShadow = cp.upperShadow;

pub fn shootingStar(self: *const CandlestickPatterns) f64 {
            if (!self.enough(2, &[_]*const CriterionState{&self.short_body, &self.long_shadow, &self.very_short_shadow})) return 0.0;

    const b1 = self.bar(2);
    const b2 = self.bar(1);

    if (!isRealBodyGapUp(b1.o, b1.c, b2.o, b2.c)) {
        return 0.0;
    }

    const mu_short = self.muLessCs(realBodyLen(b2.o, b2.c), &self.short_body, 1);
    const mu_long_us = self.muGreaterCs(upperShadow(b2.o, b2.h, b2.c), &self.long_shadow, 1);
    const mu_short_ls = self.muLessCs(lowerShadow(b2.o, b2.l, b2.c), &self.very_short_shadow, 1);

    const confidence = operators.tProductAll(&.{mu_short, mu_long_us, mu_short_ls});
    return -confidence * 100.0;
}

File: rs/src/candlestick_patterns/patterns/shooting_star.rs
//! Shooting Star pattern (2-candle bearish).

use crate::candlestick_patterns::CandlestickPatterns;
use crate::candlestick_patterns::core::{is_real_body_gap_up, lower_shadow, real_body_len, upper_shadow};
use crate::fuzzy;

/// Shooting Star: a two-candle bearish reversal pattern.
///
/// Must have:
/// - gap up from the previous candle (real body gap up),
/// - small real body,
/// - long upper shadow,
/// - very short lower shadow.
///
/// Returns:
/// Continuous float in [-100, 0].  More negative = stronger signal.
pub fn shooting_star(cp: &CandlestickPatterns) -> f64 {
    if !cp.enough(2, &[&cp.short_body, &cp.long_shadow, &cp.very_short_shadow]) {
        return 0.0;
    }

    let b1 = cp.bar(2);
    let b2 = cp.bar(1);

    if !is_real_body_gap_up(b1.o, b1.c, b2.o, b2.c) {
        return 0.0;
    }

    let mu_short = cp.mu_less(real_body_len(b2.o, b2.c), &cp.short_body, 1);
    let mu_long_us = cp.mu_greater(upper_shadow(b2.o, b2.h, b2.c), &cp.long_shadow, 1);
    let mu_short_ls = cp.mu_less(lower_shadow(b2.o, b2.l, b2.c), &cp.very_short_shadow, 1);

    let confidence = fuzzy::t_product_all(&[mu_short, mu_long_us, mu_short_ls]);
    -confidence * 100.0
}
